File: backend/app/core/websocket.py
"""
WebSocket - Real-time updates
"""
from typing import Dict, Set
from fastapi import WebSocket, WebSocketDisconnect
import json


class ConnectionManager:
    """Manages WebSocket connections per channel"""
    
    def __init__(self):
        # channel_id -> set of websockets
        self.active_connections: Dict[str, Set[WebSocket]] = {}
    
    async def connect(self, websocket: WebSocket, channel: str):
        await websocket.accept()
        if channel not in self.active_connections:
            self.active_connections[channel] = set()
        self.active_connections[channel].add(websocket)
    
    def disconnect(self, websocket: WebSocket, channel: str):
        if channel in self.active_connections:
            self.active_connections[channel].discard(websocket)
            if not self.active_connections[channel]:
                del self.active_connections[channel]
    
    async def send_to_channel(self, channel: str, message: dict):
        """Send message to all connections in a channel"""
        if channel in self.active_connections:
            disconnected = []
            for connection in self.active_connections[channel]:
                try:
                    await connection.send_json(message)
                except:
                    disconnected.append(connection)
            
            # Clean up disconnected
            for conn in disconnected:
                self.active_connections[channel].discard(conn)
    
    async def broadcast_booking_update(self, booking_id: int, status: str, data: dict = None):
        """Broadcast booking status update"""
        message = {
            "type": "booking_update",
            "booking_id": booking_id,
            "status": status,
            "data": data or {}
        }
        # Send to customer channel
        await self.send_to_channel(f"booking:{booking_id}", message)
    
    async def broadcast_worker_location(self, booking_id: int, lat: float, lng: float):
        """Broadcast worker location for tracking"""
        message = {
            "type": "worker_location",
            "booking_id": booking_id,
            "latitude": lat,
            "longitude": lng
        }
        await self.send_to_channel(f"tracking:{booking_id}", message)


# Global connection manager
manager = ConnectionManager()

File: frontend/src/services/websocket.js
/**
 * WebSocket Service for Real-time Updates
 */

class WebSocketService {
    constructor(baseUrl = 'ws://localhost:8000/api/v1/ws') {
        this.baseUrl = baseUrl;
        this.connections = new Map();
    }

    /**
     * Connect to booking updates channel
     */
    connectBookingUpdates(bookingId, token, onMessage, onError) {
        const url = `${this.baseUrl}/booking/${bookingId}?token=${token}`;
        return this._connect(`booking:${bookingId}`, url, onMessage, onError);
    }

    /**
     * Connect to worker tracking channel
     */
    connectWorkerTracking(bookingId, token, onMessage, onError) {
        const url = `${this.baseUrl}/tracking/${bookingId}?token=${token}`;
        return this._connect(`tracking:${bookingId}`, url, onMessage, onError);
    }

    /**
     * Connect as worker to send location updates
     */
    connectWorkerLocation(token, onConnected, onError) {
        const url = `${this.baseUrl}/worker/location?token=${token}`;
        const ws = new WebSocket(url);

        ws.onopen = () => {
            this.connections.set('worker:location', ws);
            if (onConnected) onConnected();
        };

        ws.onerror = (error) => {
            if (onError) onError(error);
        };

        ws.onclose = () => {
            this.connections.delete('worker:location');
        };

        return {
            sendLocation: (bookingId, lat, lng) => {
                if (ws.readyState === WebSocket.OPEN) {
                    ws.send(JSON.stringify({ booking_id: bookingId, lat, lng }));
                }
            },
            close: () => ws.close()
        };
    }

    _connect(key, url, onMessage, onError) {
        // Close existing connection if any
        if (this.connections.has(key)) {
            this.connections.get(key).close();
        }

        const ws = new WebSocket(url);

        ws.onopen = () => {
            this.connections.set(key, ws);
            // Start ping interval
            const pingInterval = setInterval(() => {
                if (ws.readyState === WebSocket.OPEN) {
                    ws.send('ping');
                } else {
                    clearInterval(pingInterval);
                }
            }, 30000);
        };

        ws.onmessage = (event) => {
            if (event.data === 'pong') return; // Ignore pong

            try {
                const data = JSON.parse(event.data);
                if (onMessage) onMessage(data);
            } catch (e) {
                console.log('WebSocket message:', event.data);
            }
        };

        ws.onerror = (error) => {
            if (onError) onError(error);
        };

        ws.onclose = () => {
            this.connections.delete(key);
        };

        return {
            close: () => ws.close(),
            isConnected: () => ws.readyState === WebSocket.OPEN
        };
    }

    /**
     * Disconnect all connections
     */
    disconnectAll() {
        this.connections.forEach((ws) => ws.close());
        this.connections.clear();
    }
}

export default new WebSocketService();
